packages/compass-connections/src/connection-scoped-global-app-registry.ts

```diff
@@ -0,0 +1,66 @@
+import type { ConnectionInfo } from '@mongodb-js/connection-info';
+import {
+  type AppRegistry,
+  createServiceLocator,
+  useGlobalAppRegistry,
+} from 'hadron-app-registry';
+import { useConnectionInfo } from '@mongodb-js/connection-storage/provider';
+import { useMemo } from 'react';
+
+type ConnectionScopedGlobalAppRegistryEvents =
+  | 'schema-analyzed'
+  | 'documents-refreshed'
+  | 'documents-paginated'
+  | 'document-inserted';
+
+export type ConnectionScopedGlobalAppRegistryLocator<
+  T extends ConnectionScopedGlobalAppRegistryEvents,
+  K extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = keyof ConnectionScopedGlobalAppRegistryImpl<T>,
+  L extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = K
+> = () => ConnectionScopedGlobalAppRegistry<T, K, L>;
+
+export type ConnectionScopedGlobalAppRegistry<
+  T extends ConnectionScopedGlobalAppRegistryEvents,
+  K extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = keyof ConnectionScopedGlobalAppRegistryImpl<T>,
+  L extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = K
+> = Pick<ConnectionScopedGlobalAppRegistryImpl<T>, K> &
+  Partial<Pick<ConnectionScopedGlobalAppRegistryImpl<T>, L>>;
+
+interface EventForwarder<T extends ConnectionScopedGlobalAppRegistryEvents> {
+  emit(event: T, payload?: Record<string, any>): void;
+}
+
+export class ConnectionScopedGlobalAppRegistryImpl<
+  T extends ConnectionScopedGlobalAppRegistryEvents
+> implements EventForwarder<T>
+{
+  constructor(
+    private readonly appRegistryEmitter: AppRegistry['emit'],
+    private readonly sourceConnectionInfo: ConnectionInfo
+  ) {}
+
+  emit(event: T, payload: Record<string, any> = {}): void {
+    payload.sourceConnectionInfo = this.sourceConnectionInfo;
+    this.appRegistryEmitter(event, payload);
+  }
+}
+
+export const connectionScopedGlobalAppRegistryLocator = createServiceLocator(
+  function useConnectionScopedGlobalAppRegistry<
+    T extends ConnectionScopedGlobalAppRegistryEvents,
+    K extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = keyof ConnectionScopedGlobalAppRegistryImpl<T>,
+    L extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = K
+  >(): ConnectionScopedGlobalAppRegistry<T, K, L> {
+    const appRegistry = useGlobalAppRegistry();
+    const connectionInfo = useConnectionInfo();
+
+    const registry = useMemo(() => {
+      return new ConnectionScopedGlobalAppRegistryImpl<T>(
+        appRegistry.emit.bind(appRegistry),
+        connectionInfo
+      );
+    }, [appRegistry, connectionInfo]);
```

Comment:
You should be careful when setting up locators like that, keep in mind that the value you get in the activate method is provided at the moment when plugin activates / mounts for the first time, so services returned from the locators should behave closer to React refs and not be memoised. In this particular case this means that you want to create your `ConnectionScopedGlobalAppRegistry` only once, and you _might_ use effects to update the state of said service, but not return memoized version of the service, this has no effect on the activate method